pysmiles/cheminfo.py

```diff
@@ -0,0 +1,182 @@
+"""
+Some simple cheminformatics tools.
+"""
+from collections import defaultdict
+import numpy as np
+import networkx as nx
+from .smiles_helper import PTE, remove_explicit_hydrogens
+
+def tanimoto_similarity(binarr1, binarr2):
+    """
+    Compute Tanimoto similarity between two binary arrays.
+
+    Parameters
+    ----------
+    binarr1: np.array
+    binarr2: np.array
+
+    Returns
+    -------
+    float
+        the tanimoto similarity
+    """
+    intersection = np.sum(binarr1 & binarr2)
+    union = np.sum(binarr1) + np.sum(binarr2) - intersection
+    return intersection / union if union != 0 else 0.0
+
+def features_to_binary(feature_list, size=1024):
+    """
+    Converte a list of features to a binary array of
+    given `size`.
+
+    Parameters
+    ----------
+    feature_list: abc.iteratable[int]
+        a iteratable with integer numbers
+    size: int
+        size of the binary array; default=1024
+
+    Returns
+    -------
+    np.array
+    """
+    fp_array = np.zeros(size, dtype=int)
+    nonzero = np.array(feature_list, dtype=int) % size
+    fp_array[nonzero] = 1
+    return fp_array
+
+def _aggregate_features(graph, cycles=4, label='fp_invariant'):
+    """
+    Given atomic invariants in a molecule stored under the `label`
+    keyword, iteratively aggregate those features for a given number
+    of `cycles`. It returns a feature list, which is composed of the
+    hashes of the aggregated features.
+
+    Parameters
+    ----------
+    graph: networkx.Graph
+    cycles: int
+        number of cycles to run; default is 4
+    label: abc.hashable
+        the keyword under which the inital invariants are stored
+
+    Returns
+    -------
+    list[int]
+        list of the hashes of unique features
+    """
+    feature_list = list(nx.get_node_attributes(graph, label).values())
+    substructures = set(nx.get_node_attributes(graph, 'subgraph').values())
+    for idx in range(0, cycles):
+        for node in graph.nodes:
+            subgraph = set(graph.nodes[node]['subgraph'])
+            feat_list_iter = [(idx, graph.nodes[node][label])]
+            for neigh in graph.neighbors(node):
+                order = graph.edges[(node, neigh)].get('order', 1)
```

Comment:
'order' needs to be a function argument. Maybe even a list of bond-features to consider. Maybe create initial bond-invariants as well